crates/stackable-operator/src/builder/pod/probe.rs

```diff
@@ -0,0 +1,330 @@
+use std::num::TryFromIntError;
+
+use k8s_openapi::{
+    api::core::v1::{ExecAction, GRPCAction, HTTPGetAction, Probe, TCPSocketAction},
+    apimachinery::pkg::util::intstr::IntOrString,
+};
+use snafu::{ResultExt, Snafu, ensure};
+
+use crate::time::Duration;
+
+#[derive(Debug, Snafu)]
+pub enum Error {
+    #[snafu(display(
+        "The probe's {field:?} duration of {duration} is too long, as it's seconds doesn't fit into an i32"
+    ))]
+    DurationTooLong {
+        source: TryFromIntError,
+        field: String,
+        duration: Duration,
+    },
+
+    #[snafu(display("The probe period is zero, but it needs to be a positive duration"))]
+    PeriodIsZero {},
+}
+
+#[derive(Debug)]
+pub struct ProbeBuilder<Action, Period> {
+    // Mandatory field
+    action: Action,
+    period: Period,
+
+    // Fields with defaults
+    success_threshold: i32,
+    failure_threshold: i32,
+    timeout: Duration,
+    initial_delay: Duration,
+    termination_grace_period: Duration,
+}
+
+impl Default for ProbeBuilder<(), ()> {
+    fn default() -> Self {
+        Self {
+            action: (),
+            period: (),
+            // The following values match the Kubernetes defaults
+            success_threshold: 1,
+            failure_threshold: 1,
+            timeout: Duration::from_secs(1),
+            initial_delay: Duration::from_secs(0),
+            termination_grace_period: Duration::from_secs(0),
+        }
+    }
+}
+
+pub enum ProbeAction {
+    Exec(ExecAction),
+    Grpc(GRPCAction),
+    HttpGet(HTTPGetAction),
+    TcpSocket(TCPSocketAction),
+}
+
+impl ProbeBuilder<(), ()> {
+    /// This probe action executes the specified command
+    pub fn with_exec_action_helper(
+        self,
+        command: impl IntoIterator<Item = impl Into<String>>,
+    ) -> ProbeBuilder<ProbeAction, ()> {
+        self.with_exec_action(ExecAction {
+            command: Some(command.into_iter().map(Into::into).collect()),
+        })
+    }
+
+    /// There is a convenience helper: [`Self::with_exec_action_helper`].
+    pub fn with_exec_action(self, exec_action: ExecAction) -> ProbeBuilder<ProbeAction, ()> {
```

Comment:
```suggestion
    /// Set an exec probe with the given configuration.
    ///
    /// For simple cases, there is a a convenience helper: [`Self::with_exec_action_helper`].
    pub fn with_exec_action(self, exec_action: ExecAction) -> ProbeBuilder<ProbeAction, ()> {
```

Reply:
I'll stop adding documenting comments, but this should apply to other pub methods too.